pkg/substitution/substitution.go

```diff
@@ -47,6 +46,33 @@ func ValidateVariable(name, value, prefix, locationName, path string, vars sets.
 
 // ValidateVariableP makes sure all variables for a parameter in the provided string are known
 func ValidateVariableP(value, prefix string, vars sets.String) *apis.FieldError {
+	if vs, present := extractVariablesWithoutValidationFromString(value, prefix); present {
+		for _, v := range vs {
+			v = strings.TrimSuffix(v, "[*]")
+
+			// contains any character except "_a-zA-Z0-9.-"?
+			re := regexp.MustCompile(`[^_a-zA-Z0-9.-]`)
+			matches := re.FindAllStringSubmatch(v, -1)
+			if len(matches) > 0 {
+				return &apis.FieldError{
+					Message: fmt.Sprintf("Invalid variable name format in value: %q. Must not contain any character other than a-z, A-Z, 0-9, or . _  -", value),
+					// Empty path is required to make the `ViaField`, … work
+					Paths: []string{""},
+				}
+			}
+
+			// starts with anything except _, a-z, A-Z?
+			re = regexp.MustCompile(`^[^_a-zA-Z].*`)
+			matches = re.FindAllStringSubmatch(v, -1)
+			if len(matches) > 0 {
+				return &apis.FieldError{
+					Message: fmt.Sprintf("Invalid variable name format in value: %q. Must not start with any character other than a-z, A-Z, 0-9, or _", value),
+					// Empty path is required to make the `ViaField`, … work
+					Paths: []string{""},
+				}
+			}
+		}
+	}
```

Comment:
Based on the three steps we discussed (pasted below), in step2, I think we don't need to do any validation here inside `ValidateVariableP`. Because validation against the param names specified in `ParamSpec` should happen inside `task_validation.go`'s `validateVariables` function, which is step 1.

> Step1: validate the format of param names at the beginning of validateVariables in `pkg/apis/pipeline/v1beta1/task_validation.go` (Implemented here https://github.com/tektoncd/pipeline/pull/4799)

> Step2: change regex to extract `NAME` for `$(params.<NAME>)` and `NAME` for `$(params["<NAME>"])` without validating anything
> - Also need to consider bracket notation in regex
> - Make sure the regex change doesn’t break others

> Step3:  Check if the extracted string is one of the param names declared in param spec. (already implemented)

Reply:
Should be ok now @chuangw6 . Let me know otherwise.